### DISCUSSION TITLE
Hide sensitive data when using with Eloquent casts.

### DISCUSSION THREAD

**Discussion by plakhin** [spatie/laravel-data#353] [Q&A]
Let's say we have a `User` model that has `credentials` attribute being casted to `CredentialsData` object and is being stored in the DB as a JSON field.
```json
{
  "service_1": {
    "id": "as9d8adn1",
    "secret": "zxcvbn"
  },
  "service_2": {
    "id": "01234",
    "secret": "qweasd"
  }
}
```

I want to hide properties like "secret" from being exposed once `User` model is being converted to an array for the response (`return auth()->user()` in e.g. `routes/api.php)`. 

 I tried to use `exceptProperties()` method like so:
 ```php
 class CredentialsData extends Data
{
    public function __construct(
        public array $service_1 = ['id' => '', 'secret' => ''],
        public array $service_2 = ['id' => '', 'secret' => ''],
    ) {
    }
    
    public function exceptProperties(): array
    {
        return [
            'service_1.secret' => true,
            'service_2.secret' => true,
        ];
    }
}
 ```
and it hides needed properties.

However, once I create a `User`:
```php
User::create([
    'email' => 'test@example.com'
    'credentials' = [
        'service_1' = ['id' => 'as9d8adn1', 'secret' => 'zxcvbn'],
        'service_2' = ['id' => '01234', 'secret' => 'qweasd'],
    ];
]);
``` 
those "secrets" are not being saved to the database.

**How can I avoid exposing those "secrets" but still be able to save them into the DB creating users as in my example.**

**Comment by rubenvanassche** [comment]
You could manually call `exclude` when sending out the data? That way the excludes are not performed on each data object. Other than that I don't see an immediate solution.

**Comment by plakhin** [comment]
@rubenvanassche,
`return $user->credentials->exclude('service_1.secret')->toArray()` has no effect, `return $user->credentials->except('service_1.secret')->toArray()` hides sensitive data.
However, the main point is to avoid accidentally exposing sensitive data by doing something like `return $user`.
And, as I wrote initially `exceptProperties()` method does this perfectly. But the problem is in this case that sensitive data is not only being hidden from output, but also is not saved to the DB.

**Comment by plakhin** [comment]
@freekmurze any chance this will be solved one day?

**Comment by plakhin** [comment]
@rubenvanassche will this be implemented in the upcoming v4?

**Comment by plakhin** [comment]
Looks like this may be solved with #505
